algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/molecular/HasMolecularDriverEventInNSCLC.kt

```diff
@@ -0,0 +1,90 @@
+package com.hartwig.actin.algo.evaluation.molecular
+
+import com.hartwig.actin.algo.evaluation.EvaluationFunction
+import com.hartwig.actin.algo.evaluation.composite.Or
+import com.hartwig.actin.algo.evaluation.util.Format
+import com.hartwig.actin.datamodel.PatientRecord
+import com.hartwig.actin.datamodel.algo.Evaluation
+import com.hartwig.actin.datamodel.algo.EvaluationResult
+import java.time.LocalDate
+
+private val ACTIVATING_MUTATION_LIST = listOf("EGFR", "ERBB2")
+private val PROTEIN_IMPACT_LIST = listOf("BRAF" to "V600E", "KRAS" to "G12C")
+private val FUSION_LIST = listOf("ALK", "NRG1", "NTRK1", "NTRK2", "NTRK3", "RET", "ROS1")
+private val EXON_SKIPPING_LIST = listOf("MET" to 14)
+private val ALL_GENES =
+    ACTIVATING_MUTATION_LIST + PROTEIN_IMPACT_LIST.map { it.first } + FUSION_LIST + EXON_SKIPPING_LIST.map { it.first }.distinct()
+
+class HasMolecularDriverEventInNSCLC(
+    private val genesToInclude: Set<String>?,
+    private val genesToIgnore: Set<String>,
+    private val maxTestAge: LocalDate? = null,
+    private val includeGenesAtLeast: Boolean,
+    private val withAvailableSOC: Boolean,
+) : MolecularEvaluationFunction(maxTestAge) {
+
+    override fun evaluate(record: PatientRecord): Evaluation {
+        return if (includeGenesAtLeast && genesToInclude != null) {
+            val evaluation = Or(createEvaluationFunctions(null, genesToIgnore)).evaluate(record)
+            if (hasPositiveEvaluationEventInGenes(evaluation, ALL_GENES - genesToInclude) && !hasPositiveEvaluationEventInGenes(
+                    evaluation,
+                    genesToInclude.toList()
+                )
+            ) {
+                clearMolecularEventsAndConfigureMessages(evaluation, true)
+            } else {
+                clearMolecularEventsAndConfigureMessages(evaluation, false)
+            }
```

Comment:
It seems redundant to check the evaluation in both calls to `hasPositive...` here. Would it make sense to pull it out? You could also factor out the call to `clearMolecular...` and just determine `mustWarn`. E.g.
```suggestion
            val mustWarn = setOf(EvaluationResult.PASS, EvaluationResult.WARN).contains(evaluation.result) &&
                hasEventForGenes(ALL_GENES - genesToInclude) && !hasEventForGenes(genesToInclude.toList())
            clearMolecularEventsAndConfigureMessages(evaluation, mustWarn)
```

Reply:
Unless this is supposed to check the individual evaluation for each gene? These are all combined with `Or` above.